components/value-distribution/src/main/java/org/datacleaner/beans/valuedist/ValueDistributionAnalyzerResultReducer.java

```diff
@@ -0,0 +1,291 @@
+            highlightedColumns[0] = groupColumn;
+            List<SingleValueDistributionResult> childResults = new ArrayList<>();
+            for (String groupName : flattenedDistinctValuesMap.keySet()) {
+                final ValueCountList reducedValueCountList = createValueCountList(reducedTopValuesMap.get(groupName));
+                SingleValueDistributionResult childValueDistributionResult = new SingleValueDistributionResult(
+                        groupName, reducedValueCountList, reducedUniqueValuesMap.get(groupName), reducedUniqueValuesMap
+                                .get(groupName).size(), reducedDistinctValuesMap.get(groupName),
+                        reducedTotalCountMap.get(groupName), annotations, new RowAnnotationImpl(),
+                        _rowAnnotationFactory, highlightedColumns);
+                childResults.add(childValueDistributionResult);
+            }
+            reducedResult = new GroupedValueDistributionResult(inputColumn, groupColumn, childResults);
+        } else {
+            final InputColumn<?>[] highlightedColumns = ((SingleValueDistributionResult) first).getHighlightedColumns();
+            final Collection<String> reducedUniqueValues = reducedUniqueValuesMap.values().iterator().next();
+            final ValueCountList reducedValueCountList = createValueCountList(reducedTopValuesMap.values().iterator()
+                    .next());
+            reducedResult = new SingleValueDistributionResult(((SingleValueDistributionResult) first).getName(),
+                    reducedValueCountList, reducedUniqueValues, reducedUniqueValues.size(), reducedDistinctValuesMap
+                            .values().iterator().next(), reducedTotalCountMap.values().iterator().next(), annotations,
+                    new RowAnnotationImpl(), _rowAnnotationFactory, highlightedColumns);
+        }
+
+        return reducedResult;
+    }
+
+    private void flattenAggregates(Map<String, Collection<ValueFrequency>> flattenedTopValuesMap,
+            Map<String, Collection<String>> flattenedUniqueValuesMap,
+            Map<String, Collection<ValueFrequency>> flattenedDistinctValuesMap,
+            Map<String, Integer> reducedTotalCountMap, final SingleValueDistributionResult singlePartialResult) {
+        Collection<ValueFrequency> flattenedTopValues = getOrInitializeFlattenedTopValues(flattenedTopValuesMap,
+                singlePartialResult);
+        Collection<String> flattenedUniqueValues = getOrInitializeFlattenedUniqueValues(flattenedUniqueValuesMap,
+                singlePartialResult);
+        Collection<ValueFrequency> flattenedDistinctValues = getOrInitializeFlattenedTopValues(
+                flattenedDistinctValuesMap, singlePartialResult);
+
+        flattenedUniqueValues.addAll(singlePartialResult.getUniqueValues());
+        flattenedTopValues.addAll(singlePartialResult.getTopValues().getValueCounts());
+        flattenedDistinctValues.addAll(singlePartialResult.getValueCounts());
```

Comment:
What I am noticing is that ```SingleValueDistributionResult.getValueCounts()``` actually contains all the data you need here ... both all the "top values", the nulls and the uniques. Would it maybe be easier to just get that list of value counts and make up all the rest from it? That way we could reduce the complexity here I think and also avoid moving the same data around into multiple collections.